compiler/rustc_codegen_ssa/src/mir/naked_asm.rs

```diff
@@ -104,6 +106,262 @@ fn inline_to_global_operand<'a, 'tcx, Bx: BuilderMethods<'a, 'tcx>>(
+        "x86_64" => (
+            if sess.target.pointer_width == 32 {
+                Architecture::X86_64_X32
+            } else {
+                Architecture::X86_64
+            },
+            None,
+        ),
+        "powerpc" => (Architecture::PowerPc, None),
+        "powerpc64" => (Architecture::PowerPc64, None),
+        "riscv32" => (Architecture::Riscv32, None),
+        "riscv64" => (Architecture::Riscv64, None),
+        "sparc" => {
+            if sess.unstable_target_features.contains(&sym::v8plus) {
+                // Target uses V8+, aka EM_SPARC32PLUS, aka 64-bit V9 but in 32-bit mode
+                (Architecture::Sparc32Plus, None)
+            } else {
+                // Target uses V7 or V8, aka EM_SPARC
+                (Architecture::Sparc, None)
+            }
+        }
+        "sparc64" => (Architecture::Sparc64, None),
+        "avr" => (Architecture::Avr, None),
+        "msp430" => (Architecture::Msp430, None),
+        "hexagon" => (Architecture::Hexagon, None),
+        "bpf" => (Architecture::Bpf, None),
+        "loongarch64" => (Architecture::LoongArch64, None),
+        "csky" => (Architecture::Csky, None),
+        "arm64ec" => (Architecture::Aarch64, Some(SubArchitecture::Arm64EC)),
+
+        // These architecutres are added here, and not present in `create_object_file`
+        "wasm32" => (Architecture::Wasm32, None),
+        "wasm64" => (Architecture::Wasm64, None),
+        "m68k" => (Architecture::M68k, None),
+
+        // Unsupported architecture.
+        _ => return None,
+    };
+
+    Some((architecture, subarchitecture))
+}
+
+/// Enable the function's target features in the body of the function, then disable them again
+fn enable_disable_target_features<'tcx>(
+    tcx: TyCtxt<'tcx>,
+    attrs: &CodegenFnAttrs,
+) -> Option<(String, String)> {
+    use std::fmt::Write;
+
+    let mut begin = String::new();
+    let mut end = String::new();
+
+    let (architecture, _subarchitecture) = parse_architecture(tcx.sess)?;
+    let features = attrs.target_features.iter().filter(|attr| !attr.implied);
+
+    match architecture {
+        Architecture::X86_64 | Architecture::X86_64_X32 | Architecture::I386 => {
+            // no action is needed, all instructions are accepted regardless of target feature
+        }
+
+        Architecture::Aarch64 | Architecture::Aarch64_Ilp32 => {
+            // https://developer.arm.com/documentation/100067/0611/armclang-Integrated-Assembler/AArch32-Target-selection-directives?lang=en
+
+            for feature in features {
+                // only enable/disable a feature if it is not already globally enabled.
+                // so that we don't infuence subsequent asm blocks
+                if !tcx.sess.unstable_target_features.contains(&feature.name) {
+                    writeln!(begin, ".arch_extension {}", feature.name).unwrap();
+
+                    writeln!(end, ".arch_extension no{}", feature.name).unwrap();
+                }
```

Comment:
This doesn't actually appear to work in some cases because due to a bug in LLVM (https://github.com/rust-lang/rust/issues/113221), even if the feature is enabled on the Rust side, it is not passed to the assembler...

However, given that the bug appears conditionally, I think it is also wrong to always set both `.arch_extension {feature}` and `.arch_extension no{feature}`.

A bit odd, but setting `.arch_extension {feature}` always and `.arch_extension no{feature}` conditionally would work in both cases.

In that approach, `.arch_extension {feature}` will affect subsequent assemblies if the feature is enabled, but what it enables is a globally enabled feature on the Rust side, so I guess there is probably no adverse effect.

Reply:
> I guess there is probably no adverse effec

Nah, this doesn't met [stateful directive rule](https://doc.rust-lang.org/nightly/reference/inline-assembly.html#r-asm.directives.stateful)...

And I think the s390x implementation of this PR that attempts to restore state based on features enabled on the Rust side has the same problem...

Reply:
I agree. We now have a separate feature flag for target features on naked functions (so that naked functions themselves can be stabilized), https://github.com/rust-lang/rust/issues/138568. At this point I'm sceptical that we can make the approach of this PR work in general. 

So, that means that either

- we use this approach on platforms where it is guaranteed to work
- we just have the user handle it (and can help out with some good documentation for that platforms that we've looked at)

Reply:
For now, I think it is safer to only support the x86/x86_64 which obviously does not require anything and architectures which support push/pop logic such as riscv, and make errors otherwise. Once the LLVM bug is fixed, we can also support aarch64 and s390x (and probably arm).

> * we just have the user handle it (and can help out with some good documentation for that platforms that we've looked at)

Considering the LLVM bug, I guess it is also difficult to use it properly on the user side.

Reply:
Whatever the design/impl outcome here, could you please make sure that is reflected in the tracking issue (#138568) itself? I'm not sure familiar with target features but the impl here makes me slightly nervous.